### ISSUE TITLE
Cyclical references in struct-valued constants?

### ISSUE THREAD

**Issue by benjamin-bader** [microsoft/thrifty#497]
This would currently break both Kotlin and (the hypothetical implementation for) Java.  Is this even allowed?  What does fbthrift do?

```thrift
struct Node {
  1: required string key;
  2: optional Foo value;
}

const Node A = {
  "key": "a",
  "value": B
}

const Node B = {
  "key": "b",
  "value": A
}
```

I can imagine using pointers in C++ for this, and builders in Java and Kotlin, but not builderless data classes...

**Comment by benjamin-bader** [comment]
Verified; this does not actually work today.  Compiling + decompiling this code reveals:

```java
// 
// Decompiled by Procyon v0.5.36
// 
public class testConstants
{
    public static final Node A;
    public static final Node B;
    
    static {
        (A = new Node()).setKey("a");
        final Node tmp0 = new Node();
        testConstants.A.setValue(tmp0);
        (B = new Node()).setKey("b");
        final Node tmp2 = new Node();
        testConstants.B.setValue(tmp2);
    }
}
```

**Comment by benjamin-bader** [comment]
Given the fact that this just silently fails in Apache thrift, I think that we can get by with failing early on cycle detection.

**Comment by benjamin-bader** [comment]
Addressed in #499 and #502
